File: backend/deployment-agent/deploy_agent/deployment_agent/supervisor.py
"""Reconcile persisted deployment runs with the real GitHub and AWS state."""

from __future__ import annotations

import threading
import time
from datetime import datetime, timedelta, timezone
from pathlib import Path
from typing import Any, Callable

from dfagents.deployer import _PERSISTED_ACTIVE_STATES, active_project_keys
from dfagents.monitor import MonitorAgent

from .models import RunState
from .state import StateStore


ORPHAN_GRACE = timedelta(minutes=20)


ANALYSIS_ORPHAN_GRACE = timedelta(minutes=90)

SUPERVISOR_INTERVAL_SECONDS = 15.0


_RECONCILED_STATES = _PERSISTED_ACTIVE_STATES | {RunState.ANALYZING.value}


def _is_orphaned(run: dict[str, Any]) -> bool:
    """True when no live in-process worker owns this run's project."""
    try:
        project_key = str(Path(run["project_path"]).resolve()).lower()
    except (OSError, ValueError):
        return False
    return project_key not in active_project_keys()


def _updated_before(run: dict[str, Any], cutoff: datetime) -> bool:
    try:
        updated = datetime.fromisoformat(str(run.get("updated_at", "")))
    except ValueError:
        return False
    if updated.tzinfo is None:
        updated = updated.replace(tzinfo=timezone.utc)
    return updated < cutoff


def reconcile_once(
    store: StateStore,
    monitor: MonitorAgent,
    emit: Callable[..., object] | None = None,
) -> list[dict[str, str]]:
    """Bring every abandoned active run back in line with reality."""
    emit = emit or (lambda *args, **kwargs: None)
    now = datetime.now(timezone.utc)
    cutoff = now - ORPHAN_GRACE
    analysis_cutoff = now - ANALYSIS_ORPHAN_GRACE
    reconciled: list[dict[str, str]] = []
    for run in store.list_runs(limit=250):
        if run["state"] not in _RECONCILED_STATES or not _is_orphaned(run):
            continue
        run_id = run["id"]
        previous = run["state"]
        repository = (run.get("repo") or {}).get("repository", "")
        if previous == RunState.ANALYZING.value:

            if not _updated_before(run, analysis_cutoff):
                continue
            store.transition_run(
                run_id,
                RunState.FAILED,
                error="Analysis worker did not survive an application restart; re-run analysis.",
            )
        elif repository:

            try:
                monitor.snapshot(run_id)
            except Exception as exc:  # noqa: BLE001 - one bad run must not stop the sweep
                emit(run_id, "warning", "supervisor", "warning", 0, f"Reconciliation deferred: {exc}")
                continue
        elif _updated_before(run, cutoff):

            store.transition_run(
                run_id,
                RunState.FAILED,
                error="Deployment worker did not survive an application restart; re-run analysis.",
            )
        else:
            continue
        current = (store.get_run(run_id) or {}).get("state", previous)
        if current == previous:
            continue
        reconciled.append({"run_id": run_id, "from": previous, "to": current})
        emit(
            run_id,
            "state",
            "supervisor",
            "complete" if current == RunState.LIVE.value else "failed"
            if current == RunState.FAILED.value
            else "running",
            100 if current in {RunState.LIVE.value, RunState.FAILED.value} else 80,
            f"Recovered abandoned {'analysis' if previous == RunState.ANALYZING.value else 'deployment'}: "
            f"{previous} to {current}",
            {"previous_state": previous, "state": current},
        )
    return reconciled


def start_supervisor(
    store: StateStore,
    monitor: MonitorAgent,
    emit: Callable[..., object] | None = None,
    interval: float = SUPERVISOR_INTERVAL_SECONDS,
) -> threading.Thread:
    """Run reconcile_once on a loop for as long as the application is alive."""

    def loop() -> None:
        while True:
            time.sleep(interval)
            try:
                reconcile_once(store, monitor, emit)
            except Exception:  # noqa: BLE001 - a transient AWS/GitHub error must not end the loop
                continue

    thread = threading.Thread(target=loop, name="deployment-supervisor", daemon=True)
    thread.start()
    return thread

File: backend/agent/deploy/supervisor.go
package deploy

import (
	"context"
	"path/filepath"
	"strings"
	"time"
)

// A run that was mid-flight when the application closed has nobody watching
// it any more. The supervisor is what notices.
//
// It cannot ask a dead worker what happened, so it asks the world instead: for
// a deployment that reached GitHub, a snapshot settles what the workflow did;
// for one that never got that far, enough time passing is the answer. Either
// way the run stops sitting in DEPLOYING for ever, which is the state a
// customer cannot do anything with.

const (
	// SupervisorInterval is how often the sweep runs. It is short because the
	// sweep is cheap for a store with no abandoned runs in it.
	SupervisorInterval = 15 * time.Second

	// orphanGrace is how long a deployment that never reached GitHub is given
	// before it is called lost.
	orphanGrace = 20 * time.Minute

	// analysisGrace is the same for an analysis, which can legitimately take
	// a long time: an install and a production build of a large project.
	analysisGrace = 90 * time.Minute
)

// reconciled are the states a run can be recovered from.
var reconciled = set(StateBootstrapping, StateCIRunning, StateDeploying, StateValidating, StateAnalyzing)

// Recovery is one run the sweep moved.
type Recovery struct {
	RunID string `json:"run_id"`
	From  string `json:"from"`
	To    string `json:"to"`
}

// Supervisor brings abandoned runs back in line with reality.
type Supervisor struct {
	Store    *Store
	Monitor  *Monitor
	Deployer *Deployer
	Emit     func(runID string, event Event)
}

// Watch reconciles on a loop until the context is cancelled. One bad run, or
// one unreachable provider, must not end the loop: the next sweep tries again.
func (s *Supervisor) Watch(ctx context.Context) {
	for {
		if !sleep(ctx, SupervisorInterval) {
			return
		}
		_, _ = s.Once(ctx)
	}
}

// Once is one sweep. It returns what it moved, which is what the Studio's
// recovery notice is built from.
func (s *Supervisor) Once(ctx context.Context) ([]Recovery, error) {
	runs, err := s.Store.ListRuns(250)
	if err != nil {
		return nil, err
	}
	now := time.Now().UTC()
	moved := []Recovery{}

	for i := range runs {
		run := &runs[i]
		if !reconciled[run.State] || s.owned(run) {
			continue
		}
		previous := run.State
		repository := text(object(run.Repo)["repository"])

		switch {
		case previous == StateAnalyzing:
			if !older(run, now.Add(-analysisGrace)) {
				continue
			}
			_, _ = s.Store.Transition(run.ID, StateFailed, map[string]any{
				"error": "Analysis worker did not survive an application restart; re-run analysis.",
			})

		case repository != "":
			// It reached GitHub, so GitHub knows what happened to it.
			if _, err := s.Monitor.Snapshot(ctx, run.ID); err != nil {
				s.emit(run.ID, Event{Type: EventLog, Stage: "supervisor", Status: StatusWarning,
					Message: "Reconciliation deferred: " + RedactText(err.Error())})
				continue
			}

		case older(run, now.Add(-orphanGrace)):
			_, _ = s.Store.Transition(run.ID, StateFailed, map[string]any{
				"error": "Deployment worker did not survive an application restart; re-run analysis.",
			})

		default:
			continue
		}

		current, _ := s.Store.GetRun(run.ID)
		if current == nil || current.State == previous {
			continue
		}
		moved = append(moved, Recovery{RunID: run.ID, From: string(previous), To: string(current.State)})
		s.announce(run.ID, previous, current.State)
	}
	return moved, nil
}

// owned reports whether a live worker in this process is still driving the
// run's project. Those are not abandoned, however long they have been going.
func (s *Supervisor) owned(run *Run) bool {
	if s.Deployer == nil {
		return false
	}
	key, err := filepath.Abs(run.ProjectPath)
	if err != nil {
		// A project path that cannot be resolved cannot be matched against a
		// worker either, so it is left alone rather than declared lost.
		return true
	}
	return s.Deployer.owns(strings.ToLower(key))
}

// older reports whether a run has not been touched since the cutoff.
func older(run *Run, cutoff time.Time) bool {
	updated, err := time.Parse(time.RFC3339Nano, run.UpdatedAt)
	if err != nil {
		return false
	}
	return updated.Before(cutoff)
}

func (s *Supervisor) emit(runID string, event Event) {
	if s.Emit != nil {
		s.Emit(runID, event)
	}
}

func (s *Supervisor) announce(runID string, from, to State) {
	what := "deployment"
	if from == StateAnalyzing {
		what = "analysis"
	}
	status, percent := StatusRunning, 80
	switch to {
	case StateLive:
		status, percent = StatusComplete, 100
	case StateFailed:
		status, percent = StatusFailed, 100
	}
	s.emit(runID, Event{Type: EventState, Stage: "supervisor", Status: status, Percent: percent,
		Message: "Recovered abandoned " + what + ": " + string(from) + " to " + string(to),
		Data:    map[string]any{"previous_state": string(from), "state": string(to)}})
}

// owns reports whether a worker in this process holds the project.
func (d *Deployer) owns(key string) bool {
	d.mu.Lock()
	defer d.mu.Unlock()
	_, taken := d.active[key]
	return taken
}
